language/resource-api/README.md

```diff
@@ -333,33 +344,166 @@ When a resource is marked with `noop => true`, either locally or through a globa
+  * `close(context)`
+    * Close the connection
+    * After calling this method the transport will not be used anymore
+    * This method should free up any caches and operating system resources (e.g. open connections)
+    * This method should never throw an exception
+
+To allow implementors a wide latitude in implementing connection and retry handling, the Resource API does not put a lot of constraints on when and how a transport can fail/error.
+
+1. Retry handling: Transports should implement all low-level retry logic necessary for dealing with network hiccups and service restarts of the target. To avoid cascading low-level issues to other parts of the system, a transport should at most wait 30 seconds for a single target to recover. When the transport has to execute retries, it is communicated to the operator at the `notice` level to raise awareness of transient problems. The transport is encouraged (but, again, not required) to activate its retry logic in the case of recoverable errors (e.g. "target unreachable").
+
+1. On `initialize`, the transport may apply deeper validation to the passed connection information, like mutual exclusivity of optional values (e.g. `password` OR `key`) and throw an `ArgumentError`
+1. On `initialize`, the transport may (but doesn't have to) try to establish a connection to the remote target. If this fails due to unrecoverable errors (e.g. "password rejected"), an appropriate exception will be thrown.
+
+1. `verify` and `facts`, like `initialize`, only needs to throw exceptions when unrecoverable errors are encountered, or when the retry logic times out.
+
+1. On any other operation will primarily be used by providers and tasks talking to this kind of target. Those operations are free to use any error signalling mechanism that is convenient. Providers and tasks will have to provide appropriate error signalling to the runtime.
+
+### Direct Access to Transports
+
+It is possible to use a RSAPI Transport directly using the `connect` method:
+
+`Puppet::ResourceApi::Transport.connect(name, config)`
+
+1. `register` the transport schema for the remote resource by:
+    * Directly calling into `Puppet::ResourceApi.register_transport`
+    * Loading an existing schema using `require 'puppet/transport/schema/<transportname>`
+    * Setting up Puppet's autoloader for `'puppet/transport/schema'`
+2. `connect` to the transport by name, passing the connection info
+3. When the transport has been initialized, `connect` will return the `Transport` object
+
+See above for possible exceptions coming from initializing a transport.
+
+To get a list of all registered transports, call the `list` method:
+
+`Puppet::ResourceApi::Transport.list`
+
+It will return a hash of all registered transport schemas keyed by their name. Each entry in the list is the transport schema definition as passed to `register_transport`.
```

Comment:
Just to be clear, does this mean each value is also a hash (exactly like the one passed to `register`)? Thinking about the API access to this, and want to figure out what kinds of data conversion we would need (or wouldn't need) to do.

Reply:
Yes. `list` will return a ruby hash of hashes. I'll need to improve the wording on this one a little bit to make that clearer.

The current implementation is quite primitive: https://github.com/puppetlabs/puppet-resource_api/pull/150/commits/bd3a21e0abd825f55ef928189d5ebb8a256b7c6c